File: app/src/main/java/com/credenceid/sdkapp/FaceActivity.java
package com.credenceid.sdkapp;

import android.app.Activity;
import android.app.AlertDialog;
import android.content.Intent;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.PointF;
import android.graphics.RectF;
import android.os.Build;
import android.os.Bundle;

import com.credenceid.biometrics.Biometrics;
import com.credenceid.face.FaceEngine;
import com.credenceid.sdkapp.util.BitmapUtils;
import com.credenceid.util.FileUtils;

import java.util.ArrayList;

import static com.credenceid.biometrics.Biometrics.ResultCode.FAIL;
import static com.credenceid.biometrics.Biometrics.ResultCode.INTERMEDIATE;
import static com.credenceid.biometrics.Biometrics.ResultCode.OK;
import static com.credenceid.biometrics.DeviceFamily.CredenceTwo;

@SuppressWarnings("StatementWithEmptyBody")
public class FaceActivity
	extends Activity {

	private AlertDialog.Builder mBuilder;
	private AlertDialog mDialog;

	@Override
	protected void
	onCreate(Bundle savedInstanceState) {

		super.onCreate(savedInstanceState);
		setContentView(R.layout.act_face);

		/* Start by displaying a popup to let user known a background operation is taking place. */
		mBuilder = new AlertDialog.Builder(this);
		mBuilder.setCancelable(false);
		mBuilder.setIcon(R.drawable.ic_launcher);
		mBuilder.setTitle(getString(R.string.face_engine_analytics_title));
		mBuilder.setMessage(getString(R.string.face_engine_analyzing_wait));

		mDialog = mBuilder.create();
		mDialog.setCanceledOnTouchOutside(false);
		mDialog.show();

		this.readAndAnalyzeImage();
	}

	private void
	readAndAnalyzeImage() {

		Bitmap image = null;

		/* There is an Android framework's bug in Oreo where sending image data through
		 * Intent's causes an "FAILED BINDER TRANSACTION" crash to occur. This same block of
		 * code works as expected on Marshmallow, Nougat, Pie, yet fails on ALL Android Oreo
		 * devices. To get around this in camera.Activity we save image then pass image file path.
		 */
		if (Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.O)
			image = FileUtils.readBitmap(getIntent().getStringExtra("image"));
		else {
			byte[] imageBytes = getIntent().getByteArrayExtra("image");
			if (null != imageBytes)
				image = BitmapFactory.decodeByteArray(imageBytes, 0, imageBytes.length);
		}

		if (null != image) {
			if (CredenceTwo == App.DevFamily)
				image = BitmapUtils.rotate(image, 90);

			this.analyzeImage(image);
		} else {
			mDialog.dismiss();
			mBuilder.setTitle(getString(R.string.error));
			mBuilder.setMessage(getString(R.string.err_cant_get_image));
			mBuilder.setPositiveButton(getString(R.string.retry), (dialogInterface, i) -> {
				finish();
				startActivity(new Intent(getApplicationContext(), CameraActivity.class));
			});
			mDialog = mBuilder.create();
			mDialog.show();
		}
	}

	/* Performs a full face analysis (via CredenceSDK) on given image.
	 *
	 * @param bitmap Image to run full face analysis on.
	 */
	@SuppressWarnings("StringConcatenationInLoop")
	private void
	analyzeImage(final Bitmap bitmap) {

		if (null == bitmap)
			return;

		/* Make API call to run full face analysis. */
		App.BioManager.analyzeFace(bitmap, (Biometrics.ResultCode resultCode,
		                                    RectF rectF,
		                                    ArrayList<PointF> arrayList,
		                                    ArrayList<PointF> arrayList1,
		                                    float[] floats,
		                                    FaceEngine.HeadPoseDirection[] poseDirections,
		                                    FaceEngine.Gender gender,
		                                    int age,
		                                    FaceEngine.Emotion emotion,
		                                    boolean glasses,
		                                    int imageQuality) -> {

			String message = "";

			if (OK == resultCode) {
				String displayData = "HeadPose: ";
				for (FaceEngine.HeadPoseDirection pose : poseDirections)
					displayData += (pose.name() + " ");

				displayData += ("\nGender: " + gender.name());
				displayData += ("\nAge: " + age);
				displayData += ("\nEmotion: " + emotion.name());
				displayData += ("\nImage Quality: " + imageQuality);

				message = displayData;

			} else if (INTERMEDIATE == resultCode) {
				/* This code is never returned here. */

			} else if (FAIL == resultCode) {
				message = getString(R.string.face_engine_fail);
			}

			/* Remove popup, update its message with result from API call, and re-display. */
			mDialog.dismiss();
			mBuilder.setMessage(message);
			mBuilder.setPositiveButton(getString(R.string.retry), (dialogInterface, i) -> {
				finish();
				startActivity(new Intent(getApplicationContext(), CameraActivity.class));
			});
			mDialog = mBuilder.create();
			mDialog.show();
		});
	}
}

File: app/src/main/java/com/credenceid/sdkapp/FaceActivity.kt
@file:Suppress("DEPRECATION")

package com.credenceid.sdkapp

import android.annotation.SuppressLint
import android.app.Activity
import android.app.ProgressDialog
import android.content.Intent
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.PointF
import android.graphics.RectF
import android.os.Bundle
import android.os.Handler
import android.widget.Toast
import android.widget.Toast.LENGTH_LONG
import com.credenceid.biometrics.Biometrics
import com.credenceid.biometrics.Biometrics.ResultCode.*
import com.credenceid.biometrics.DeviceFamily.CredenceTwo
import com.credenceid.face.FaceEngine
import com.credenceid.sdkapp.android.camera.Utils
import kotlinx.android.synthetic.main.act_face.*
import java.util.*

@Suppress("DEPRECATION")
class FaceActivity : Activity() {

    private lateinit var dialog: ProgressDialog

    override fun onCreate(savedInstanceState: Bundle?) {

        super.onCreate(savedInstanceState)
        setContentView(R.layout.act_face)

        dialog = ProgressDialog(this)
        dialog.setMessage(getString(R.string.processing))
        dialog.setProgressStyle(ProgressDialog.STYLE_SPINNER)
        dialog.isIndeterminate = true
        dialog.setCancelable(false)

        this.configureLayoutComponents()

        /* If bytes were given to this activity, perform face analysis. */
        val imageBytes = intent.getByteArrayExtra(getString(R.string.camera_image))
        if (null != imageBytes) {
            var image = BitmapFactory.decodeByteArray(imageBytes, 0, imageBytes.size)
            if (CredenceTwo == App.DevFamily)
                image = Utils.rotateBitmap(image, 90f)

            this.detectFace(image)
        } else {
            this.onBackPressed()

            Handler().postDelayed({
                Toast.makeText(this, "Failed to detect face.", LENGTH_LONG).show()
            }, 1000)
        }
    }

    override fun onBackPressed() {

        super.onBackPressed()
        this.finish()
    }

    private fun configureLayoutComponents() {

        finishBtn.setOnClickListener { this.finish() }
    }

    @SuppressLint("SetTextI18n")
    fun detectFace(bitmap: Bitmap?) {

        /* If invalid Bitmap, display Toast and exit out.*/
        if (null == bitmap) {
            Toast.makeText(this, getString(R.string.no_image_found_to_process), LENGTH_LONG).show()
            return
        }

        /* Display dialog so user knows an operation is in progress. */
        this.showProgressDialog()

        /* Create new scaled image to run analysis on. */
        faceImageView.setImageBitmap(bitmap)

        App.BioManager!!.analyzeFace(bitmap) { rc: Biometrics.ResultCode,
                                               _: RectF,
                                               _: ArrayList<PointF>,
                                               _: ArrayList<PointF>,
                                               _: FloatArray,
                                               poseDir: Array<FaceEngine.HeadPoseDirection>,
                                               gender: FaceEngine.Gender,
                                               age: Int,
                                               emotion: FaceEngine.Emotion,
                                               glasses: Boolean,
                                               imageQuality: Int ->

            /* If we got back data, populate CropView and other widgets with face data. */
            when (rc) {
                OK -> {
                    genderTextView.text = getString(R.string.gender_colon_arg) + gender.name
                    ageTextView.text = getString(R.string.age_colon_arg) + age
                    glassesTextView.text = getString(R.string.glasses_colon_arg) + glasses
                    emotionTextView.text = getString(R.string.emotiona_colon_arg) + emotion.name
                    qualityTextView.text = getString(R.string.imagequal_colon_arg) + imageQuality + "%"

                    var text = getString(R.string.headposedir_colon_arg)

                    text += if (poseDir[1] == FaceEngine.HeadPoseDirection.STRAIGHT) {
                        if (poseDir[2] == FaceEngine.HeadPoseDirection.STRAIGHT)
                            "STRAIGHT"
                        else
                            poseDir[2].name

                    } else if (poseDir[2] == FaceEngine.HeadPoseDirection.STRAIGHT)
                        poseDir[1].name
                    else
                        poseDir[1].name + "\n  &\n" + poseDir[2].name

                    poseDirTextView.text = text

                }
                INTERMEDIATE -> {
                    /* This code is never returned for this API. */
                }
                FAIL -> {
                    this.onBackPressed()
                    Handler().postDelayed({
                        Toast.makeText(this, "Failed to detect face.", LENGTH_LONG).show()
                    }, 1000)
                }
                else -> {
                }
            }

            this.dismissProgressDialog()
        }
    }

    private fun showProgressDialog() {

        if (!dialog.isShowing)
            dialog.show()
    }

    private fun dismissProgressDialog() {

        if (dialog.isShowing)
            dialog.dismiss()
    }

}